### PR TITLE
Refactored Nag Dialogs, Split `UnableToAffordExpensesNagDialog` into Two Dialogs, Added Unit Tests

### LINKED ISSUE #4834: Unable to afford expenses nag appears when current funds are below loan collateral.
### Prerequisites and Pre-Issue Checklist

- [X] I'm reporting the issue to the correct repository:
- [MegaMek](https://github.com/MegaMek/megamek/issues)
- [MegaMekLab](https://github.com/MegaMek/megameklab/issues)
- [MekHQ](https://github.com/MegaMek/mekhq/issues)

- [X] I've tested the issue against at least the latest MILESTONE version
- [X] I've asked on the MegaMek Discord about the error
- [X] I've reviewed the BattleTech rules and MekHQ documentation, and I've confirmed that something isn't working as intended.
- [X] I've searched the Github tracker and haven't found the issue listed

### Severity *

Low (Minor/Nuisance): Minor glitches or cosmetic issues that don’t affect gameplay and occur rarely.

### Brief Description *

When taking out a loan and having total funds below the collateral of the loan causes a nag message to appear at the end of the month. This nag message claims the inability to cover all expenses, even if this is strictly speaking not true. 
E.g. one can be below a 10mil collateral, and still be quite capable of paying expenses, while this nag suggests imminent bankrupcy instead. 

image

This is possibly functioning as intended, but I was requested to bring this up just for a sanity check. 

### Steps to Reproduce

1. Take out a loan through the finances tab. 
2. Take note of agreed collateral.
3. Ensure current funds are below the agreed collateral (GM adding a negative amount worked fine on my end). 
4. Proceed to end of month, observe nag message. 

### Operating System *

Windows 11

### Java Version *

21.0.3

### MekHQ Suite Version *

Free Text (type manually)

### Custom MekHQ Version

0.50.01-SNAPSHOT - Revision: e0a790a6a4cfccb36f50625799b23898e0e8bd1a

### Attach Files

_No response_

### Final Checklist

- [X] I've checked to make sure that this issue has not already been filed
- [X] I'm reporting only one issue in this ticket for clarity and focus

### PR DESCRIPTION
**Pull request by IllianiCBT**
Refactored a majority of our existing nag dialogs to standardize the code used within.

Modified the `UnableToAffordExpensesNagDialog` to specify that only monthly expenses are being checked. Updated the Java method to exclude loan data.

Added a new nag that triggers specifically if the campaign is unable to afford any loan payments due the next day.

Moved the StratCon outstanding contact nag logic into the relevant nag dialog class.

Added Unit Tests for every nag dialog, barring the single dialog that already had Unit Tests.

### Closes #4834

### CODE REVIEW

**Comment by IllianiCBT** [comment]
Flipping to draft while I write some Unit Tests

File: MekHQ/src/mekhq/gui/CampaignGUI.java
Comment (IllianiCBT) [inline MekHQ/src/mekhq/gui/CampaignGUI.java:2542]: Moved these functions into their own discrete methods, as it made sense to isolate them in case we ever want to move them into their own class (like the other nag dialogs).

File: MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordExpensesNagDialog.java
Comment (rjhancock) [inline MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordExpensesNagDialog.java:31]: Don't forget a JavaDoc for the class itself.

File: MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordLoanPaymentNagDialog.java
Comment (rjhancock) [inline MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordLoanPaymentNagDialog.java:2]: Need to correct the year here.

File: MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordLoanPaymentNagDialog.java
Comment (rjhancock) [inline MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordLoanPaymentNagDialog.java:32]: Class level JavaDoc.

File: MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordLoanPaymentNagDialog.java
Comment (rjhancock) [inline MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordLoanPaymentNagDialog.java:67]: Turn these strings into constants.

File: MekHQ/unittests/mekhq/campaign/event/DayEndingEventTest.java
Comment (rjhancock) [inline MekHQ/unittests/mekhq/campaign/event/DayEndingEventTest.java:1]: Dont' forget the correct copyright comments.

File: MekHQ/unittests/mekhq/campaign/event/DayEndingEventTest.java
Comment (rjhancock) [inline MekHQ/unittests/mekhq/campaign/event/DayEndingEventTest.java:9]: Class JavaDoc

File: MekHQ/unittests/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialogTest.java
Comment (rjhancock) [inline MekHQ/unittests/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialogTest.java:1]: Copyright Comment Block.

File: MekHQ/unittests/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialogTest.java
Comment (rjhancock) [inline MekHQ/unittests/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialogTest.java:63]: Make a helper method that will generate the units and return a list of Mocked units so you only need to do the code once for all tests.

Can take 4 parameters for maintained/salvage for each unit and just generate what it needs and returns a list of Unit's that are mocked.

Comment (rjhancock) [reply MekHQ/unittests/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialogTest.java:63]: Can have it generate the campaign object as well to make it a one call method.

File: MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordLoanPaymentNagDialog.java
Comment (IllianiCBT) [reply MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordLoanPaymentNagDialog.java:67]: image

I get this error when attempting to use constants stored at the beginning of the Class

Comment (rjhancock) [reply MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordLoanPaymentNagDialog.java:67]: Are they static?

Comment (IllianiCBT) [reply MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordLoanPaymentNagDialog.java:67]: Derp

Comment (rjhancock) [reply MekHQ/src/mekhq/gui/dialog/nagDialogs/UnableToAffordLoanPaymentNagDialog.java:67]: Don't beat yourself up over it. You're still learning.

File: MekHQ/src/mekhq/gui/CampaignGUI.java
Comment (Sleet01) [reply MekHQ/src/mekhq/gui/CampaignGUI.java:2542]: Just an FYI, not requesting any changes:

Private members of a class are only visible to instances of that class, so will not be available for unit testing.  
It is possible to write less-protected wrapper functions that then call the `private` methods just for testing purposes - that is, they can then be called in unit tests, but the private methods themselves will be invisible to unit testing code.

For that reason, it is common to see code that has unit tests get switched from `private` to `protected` - the Junit unit test code needs at least this much access to call the methods.
Package-private is another option _iff_ the unit tests can be placed in the package hierarchy of the  code under test, but I don't think we have that option with Junit 5.

File: MekHQ/unittests/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialogTest.java
Comment (Sleet01) [inline MekHQ/unittests/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialogTest.java:132]: Might be a typo: `noUoUnmaintainedUnitExistsUnit1()` has an extraneous `Uo`.

File: MekHQ/unittests/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialogTest.java
Comment (Sleet01) [inline MekHQ/unittests/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialogTest.java:102]: This is all really good!  I have a single recommendation, not necessary to implement here but something to think about going forward:

Check out "parameterization", a method for reducing copy-and-pasting a bunch of similar tests (q.v. https://www.baeldung.com/parameterized-tests-junit-5).
When you have a set of tests that are identical except for some parameters that they consume, you can have Junit run the same test function with different combinations of parameters, thusly:

```
@ParameterizedTest
@ValueSource(ints = {1, 3, 5, -3, 15, Integer.MAX_VALUE})
void isOdd_ShouldReturnTrueForOddNumbers(int number) {
    assertTrue(Numbers.isOdd(number));
}
```

So this allows for running a test function multiple times with different inputs, cutting down on repetitive code.
There are drawbacks, though:

> One of the limitations of value sources is that they only support these types:
> short (with the shorts attribute)
> byte (bytes attribute)
> int (ints attribute)
> long (longs attribute)
> float (floats attribute)
> double (doubles attribute)
> char (chars attribute)
> java.lang.String (strings attribute)
> java.lang.Class (classes attribute)
> Also, we can only pass one argument to the test method each time.

Other things to note:
This only works if the assert is the same for every combo of parameters _or_ if you can pass in the expected value as an additional parameter... which is usually too much hassle.
So in _this_ case, you could parameterize all the `assertTrue` tests into one function, and all the `assertFalse` tests into another, if `initializeUnits()` could be set up to take a single value rather than 4.

Again, not requesting any changes here, just something you might want to look at for future work.

**Review by Sleet01 — APPROVED** [review]
Looks very good, outside of already-identified changes and one minor typo.

File: MekHQ/unittests/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialogTest.java
Comment (IllianiCBT) [reply MekHQ/unittests/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialogTest.java:102]: This is very cool, thank you for the reading suggestion. I might be able to get it to work here if I can work out an algorithm that will allow us to test all the possible combinations, but I think in this instance what we would gain in brevity we'd lose in readability.

I know you said you weren't suggesting change, but I wanted to comment my thought process.

File: MekHQ/src/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialog.java
Comment (rjhancock) [inline MekHQ/src/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialog.java:56]: Use the JavaDoc code for linking to the Campaign class.

File: MekHQ/unittests/mekhq/gui/dialog/nagDialogs/EndContractNagDialogTest.java
Comment (rjhancock) [inline MekHQ/unittests/mekhq/gui/dialog/nagDialogs/EndContractNagDialogTest.java:1]: Copyright....

Comment (IllianiCBT) [reply MekHQ/unittests/mekhq/gui/dialog/nagDialogs/EndContractNagDialogTest.java:1]: >_>

File: MekHQ/src/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialog.java
Comment (IllianiCBT) [reply MekHQ/src/mekhq/gui/dialog/nagDialogs/UnmaintainedUnitsNagDialog.java]: I had no idea {@link} existed, thanks for the tip!

**Comment by IllianiCBT** [comment]
Apologies to anyone following this PR. I'm trying one last thing to see whether I can work around the headless environment issue, before I go to the effort of extracting Windchild's nag code into its own static method.

**Comment by rjhancock** [comment]
It might require a change to the build scripts to support headless mode. But no logic should actually BE in any view, it should all be in data style classes for testing and the view should just need to extract what it needs.

**Comment by IllianiCBT** [comment]
Yeah, I suspect you're right, but Google suggested trying `System.setProperty("java.awt.headless", "true")`, so for the sake of avoiding extra work I figured it was worth the try (the attempt failed, so at this point I'm just going to extract the nag trigger logic into its own method).

**Comment by IllianiCBT** [comment]
Flipping this to 'live'. Description and title have been updated to reflect changes.

@Sleet01 there have been a bunch more Unit Tests added since your earlier review. I don't know if you want to take another look before confirming it has your seal of approval.

**Comment by rjhancock** [comment]
Other than the JavaDoc stuff I mentioned, (yes I know, minor stuff but we all need to get better at it), it looks good to me.

**Comment by IllianiCBT** [comment]
That should all be handled now - was there one I missed?

**Comment by rjhancock** [comment]
Half a dozen or so. I tagged them in the review changes section.

**Comment by IllianiCBT** [comment]
I _believe_ I've caught the last remaining JavaDoc issues.

**Comment by rjhancock** [comment]
To help find the issues, run `./gradlew javadoc` and it'll output a report of warnings and errors. Errors MUST be fixed, warnings RELATED to your code SHOULD be but NOT required.

**Comment by IllianiCBT** [comment]
That command is invaluable, thank you

**Comment by rjhancock** [comment]
I'll be adding that to the Actions soon as part of a requirement of things passing. And eventually doing one for GH Pages of JavaDoc so be mindful and descriptive of your documentation. Write it for people who DON'T code.

**Comment by IllianiCBT** [comment]
Sounds good. I've been getting into the habit of writing JavaDocs for just about everything.

**Pull request merged**
